Dockerfile

```diff
@@ -245,32 +249,31 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
 
 # dockercli
 FROM base AS dockercli-src
-WORKDIR /tmp/dockercli
-RUN git init . && git remote add origin "https://github.com/docker/cli.git"
-ARG DOCKERCLI_VERSION
-RUN git fetch -q --depth 1 origin "${DOCKERCLI_VERSION}" +refs/tags/*:refs/tags/* && git checkout -q FETCH_HEAD
-RUN [ -d ./components/cli ] && mv ./components/cli /usr/src/dockercli || mv /tmp/dockercli /usr/src/dockercli
 WORKDIR /usr/src/dockercli
+COPY hack/dockerfile/cli.sh ./build.sh
+ARG DOCKERCLI_REPOSITORY
+RUN git init . && git remote add origin ${DOCKERCLI_REPOSITORY}
+ARG DOCKERCLI_VERSION
+ARG DOCKERCLI_INTEGRATION_VERSION
+RUN git fetch -q --depth 1 origin "${DOCKERCLI_INTEGRATION_VERSION}" "${DOCKERCLI_VERSION}" +refs/tags/*:refs/tags/*
```

Comment:
I see that you're trying to optimize by deduplicating the git fetch, but I suspect it is actually going to be suboptimal:
- Changing either `DOCKERCLI_INTEGRATION_VERSION` or `DOCKERCLI_VERSION` invalidates the cache for both CLI versions
- In the common case (tagged releases) the repository is going to be wholly unused as `cli.sh` downloads the binaries from download.docker.com

This construct is going to be optimal only in the case where both versions of the CLI need to be built from source. I think it would be faster most of the time, especially interactively, to have `cli.sh` pull the source only after downloading the prebuilt binary fails.

Reply:
Good point! Moved the fetch to the `cli.sh`.